graphbolt/src/cuda/common.h

```diff
@@ -82,15 +83,34 @@ inline bool is_zero<dim3>(dim3 size) {
 
 #define CUDA_CALL(func) C10_CUDA_CHECK((func))
 
-#define CUDA_KERNEL_CALL(kernel, nblks, nthrs, shmem, stream, ...)    \
-  {                                                                   \
-    if (!graphbolt::cuda::is_zero((nblks)) &&                         \
-        !graphbolt::cuda::is_zero((nthrs))) {                         \
-      (kernel)<<<(nblks), (nthrs), (shmem), (stream)>>>(__VA_ARGS__); \
-      C10_CUDA_KERNEL_LAUNCH_CHECK();                                 \
-    }                                                                 \
+#define CUDA_KERNEL_CALL(kernel, nblks, nthrs, shmem, ...)          \
+  {                                                                 \
+    if (!graphbolt::cuda::is_zero((nblks)) &&                       \
+        !graphbolt::cuda::is_zero((nthrs))) {                       \
+      auto stream = graphbolt::cuda::GetCurrentStream();            \
+      (kernel)<<<(nblks), (nthrs), (shmem), stream>>>(__VA_ARGS__); \
+      C10_CUDA_KERNEL_LAUNCH_CHECK();                               \
+    }                                                               \
   }
 
+#define CUB_CALL(fn, ...)                                                     \
+  {                                                                           \
+    auto allocator = graphbolt::cuda::GetAllocator();                         \
+    auto stream = graphbolt::cuda::GetCurrentStream();                        \
+    size_t workspace_size = 0;                                                \
+    CUDA_CALL(cub::fn(nullptr, workspace_size, __VA_ARGS__, stream));         \
+    auto workspace = allocator.AllocateStorage<char>(workspace_size);         \
+    CUDA_CALL(cub::fn(workspace.get(), workspace_size, __VA_ARGS__, stream)); \
+  }
+
+#define THRUST_CALL(fn, ...)                                                 \
+  [&] {                                                                      \
```

Comment:
wrap with `{}` and end with `;` before `}`?

Reply:
This is like a function call and we expect the users of this macro to end their statements with `;` of their own. If we add `;` at the end, then the user can finish their lines with no `;` at the end of `THRUST_CALL`:
```
int b = f(3);
THRUST_CALL(...)
int a = f(2);
```
But we want:
```
int b = f(3);
THRUST_CALL(...);
int a = f(2);
```

Reply:
I needed to use `lambda` to be able to handle return values from thrust functions.

Reply:
I hope we are on the same page, if not you can suggest a change here and I will make sure to fix it in a future PR immediately. Merging so that I can adapt other PRs on top of this PR.